### PR TITLE
AP-1011 archive_load_files feature

### PR DESCRIPTION
**Pull request by llehtinen**
## Problem

https://transferwise.atlassian.net/browse/AP-1011

## Proposed changes

When enabled, this feature will result in PipelineWise load files being copied to a separate folder for archival.

When used in conjunction with incremental replication, the replication key and its min/max value in each archived file are included in S3 metadata.

Original grooming document (slightly outdated now):
https://docs.google.com/document/d/11UTlmWVJS9aGickmyxpXOUhrv4RTPnm8zip_IqeR2J0/edit?ts=609e4106

(Old WIP PR: https://github.com/transferwise/pipelinewise-target-snowflake/pull/174)

## Types of changes

What types of changes does your code introduce to PipelineWise?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [ ] Description above provides context of the change
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Unit tests for changes (not needed for documentation changes)
- [ ] CI checks pass with my changes
- [ ] Bumping version in `setup.py` is an individual PR and not mixed with feature or bugfix PRs
- [ ] Commit message/PR title starts with `[AP-NNNN]` (if applicable. AP-NNNN = JIRA ID)
- [ ] Branch name starts with `AP-NNN` (if applicable. AP-NNN = JIRA ID)
- [ ] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [ ] Relevant documentation is updated including usage instructions

### CODE REVIEW

**Review by koszti — APPROVED** [review]

File: target_snowflake/__init__.py [hunk 1]
```diff
@@ -113,6 +114,8 @@ def persist_lines(config, lines, table_cache=None, file_format_type: FileFormatT
     batch_size_rows = config.get('batch_size_rows', DEFAULT_BATCH_SIZE_ROWS)
     batch_wait_limit_seconds = config.get('batch_wait_limit_seconds', None)
     flush_timestamp = datetime.utcnow()
+    archive_load_files_enabled = config.get('archive_load_files', {}).get('enabled', None)
```
Comment (koszti) [inline target_snowflake/__init__.py:117]:
```suggestion
    archive_load_files = config.get('archive_load_files')
```

File: target_snowflake/__init__.py [hunk 2]
```diff
@@ -170,6 +173,21 @@ def persist_lines(config, lines, table_cache=None, file_format_type: FileFormatT
             else:
                 records_to_load[stream][primary_key_string] = o['record']
 
+            if archive_load_files_enabled and stream in archive_load_files_data:
```
Comment (koszti) [inline target_snowflake/__init__.py:176]:
```suggestion
            if archive_load_files and stream in archive_load_files_data:
```

can we change it to be single level option?

File: target_snowflake/__init__.py [hunk 3]
```diff
@@ -257,6 +277,27 @@ def persist_lines(config, lines, table_cache=None, file_format_type: FileFormatT
                 else:
                     stream_to_sync[stream] = DbSync(config, o, table_cache, file_format_type)
 
+                if archive_load_files_enabled:
```
Comment (koszti) [inline target_snowflake/__init__.py:280]:
```suggestion
                if archive_load_files:
```

can we change it to be single level option?

File: target_snowflake/db_sync.py [hunk 4]
```diff
@@ -63,6 +63,11 @@ def validate_config(config):
     if not config_default_target_schema and not config_schema_mapping:
         errors.append("Neither 'default_target_schema' (string) nor 'schema_mapping' (object) keys set in config.")
 
+    # Check if archive load files option is using external stages
+    archive_load_files = config.get('archive_load_files', {})
+    if archive_load_files.get('enabled') and not config.get('s3_bucket', None):
```
Comment (koszti) [inline target_snowflake/db_sync.py:68]:
```suggestion
    if archive_load_files and not config.get('s3_bucket', None):
```

File: tests/integration/test_target_snowflake.py [hunk 5]
```diff
@@ -1159,3 +1176,59 @@ def test_parquet(self):
 
         # Check if data loaded correctly and metadata columns exist
         self.assert_three_streams_are_into_snowflake()
+
+    def test_archive_load_files(self):
+        """Test if load file is copied to archive folder"""
+        self.config['archive_load_files'] = {'enabled': True}
```
Comment (koszti) [inline tests/integration/test_target_snowflake.py:1182]:
```suggestion
        self.config['archive_load_files'] = True
```

File: tests/unit/test_db_sync.py [hunk 6]
```diff
@@ -86,6 +86,11 @@ def test_config_validation(self):
         config_with_external_stage['stage'] = 'dummy-value'
         self.assertGreater(len(validator(config_with_external_stage)), 0)
 
+        # Configuration with archive_load_files but no s3_bucket
+        config_with_archive_load_files = minimal_config.copy()
+        config_with_archive_load_files['archive_load_files'] = {'enabled': True}
```
Comment (koszti) [inline tests/unit/test_db_sync.py:91]:
```suggestion
        config_with_archive_load_files['archive_load_files'] = True
```

File: tests/unit/test_target_snowflake.py [hunk 7]
```diff
@@ -85,4 +85,61 @@ def test_persist_40_records_with_batch_wait_limit(self, dbSync_mock, flush_strea
...
         # Expecting flush after every records + 1 at the end
-        assert flush_streams_mock.call_count == 41
\ No newline at end of file
+        assert flush_streams_mock.call_count == 41
+
+    @patch('target_snowflake.DbSync')
+    @patch('target_snowflake.os.remove')
+    def test_archive_load_files_incremental_replication(self, os_remove_mock, dbSync_mock):
+        self.config['tap_id'] = 'test_tap_id'
+        self.config['archive_load_files'] = {'enabled': True}
```
Comment (koszti) [inline tests/unit/test_target_snowflake.py:94]:
```suggestion
        self.config['archive_load_files'] = True
```

File: tests/unit/test_target_snowflake.py [hunk 8]
```diff
@@ -85,4 +85,61 @@ def test_persist_40_records_with_batch_wait_limit(self, dbSync_mock, flush_strea
...
+            'archive-load-files-primary-column': 'id',
+            'archive-load-files-primary-column-min': '1',
+            'archive-load-files-primary-column-max': '5'
+        }
+
+    @patch('target_snowflake.DbSync')
+    @patch('target_snowflake.os.remove')
+    def test_archive_load_files_log_based_replication(self, os_remove_mock, dbSync_mock):
+        self.config['tap_id'] = 'test_tap_id'
+        self.config['archive_load_files'] = {'enabled': True}
```
Comment (koszti) [inline tests/unit/test_target_snowflake.py:124]:
```suggestion
        self.config['archive_load_files'] = True
```

**Review by koszti — CHANGES_REQUESTED** [review]
Can you please change the multi level `archive_load_files.enabled` option to be simply `archive_load_files`? The options in `config.json` are always single level and keeping it consistent would make life easier. I added suggestions to the PR above.

Also, can you please document the new `archive_load_files` optional parameter in `README.md`?

**Comment by llehtinen** [comment]
@koszti Requested changes have been made. In addition, as discussed on Slack, the `archive-load-files-primary-column` term was replaced with `incremental-key` in metadata, variable names, and `stream_utils` method name.

**Review by koszti — APPROVED** [review]

**Pull request merged**